libevmasm/SemanticInformation.h

```diff
@@ -45,6 +45,32 @@ struct SemanticInformation
 		Write
 	};
 
+	enum class Location { Storage, Memory };
+
+	/**
+	 * Represents a read or write operation from or to one of the data locations.
+	 */
+	struct Operation
+	{
+		Location location;
+		Effect effect;
+		/// Start of affected area as an index into the parameters.
+		/// Unknown if not provided.
+		std::optional<size_t> startParameter;
+		/// Length of the affected area as an index into the parameters (if this is an opcode).
+		/// Unknown if neither this nor lengthConstant is provided.
+		std::optional<size_t> lengthParameter;
+		/// Length as a constant.
+		/// Unknown if neither this nor lengthArgument is provided.
+		std::optional<size_t> lengthConstant;
+	};
+
+	/// @returns the sequence of read write operations performed by the instruction.
+	/// Order matters.
+	/// For external calls, there is just one unknown read and one unknown write operation,
+	/// event though there might be multiple.
+	static std::vector<Operation> readWriteOperations(Instruction _instruction);
```

Comment:
How about something like `readWriteOperations` that takes a `FunctionCall` (always builtin) rather than just an instruction. I think the returned object should be more expressive. Maybe a variant `Memory(Expression& offset, Expression& length)` and `Storage(Expression& slot)`.

Reply:
This is inside libevmasm. I was considering putting it into the scope of yul, but in the end this was easier. We can extend it on the yul side to full function side effects if we want, but then it would be a different structure.